### PR TITLE
Implement secio

### PR DESCRIPTION
**Pull request by tomaka**
Ready for review!

Successfully tested against both the Go and JS implementations.

Two contentious points that still exist:

- The `ring` library doesn't like some stuff inside DER public keys. Therefore I have to remove the 24 first bytes from public keys generated by OpenSSL and sent by remote implementations in order to make it work. This is obviously not a great solution, but writing a DER parser is not a trivial task.

- Right now this only accepts RSA keys, because the Go and JS implementations only use RSA, despite the fact that the protocol also accepts Ed25519 and Secp256k1.

### CODE REVIEW

File: libp2p-secio/src/handshake.rs [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+                    }
+                })
+        });
+
+	Box::new(future)
+}
+
+// Custom algorithm translated from reference implementations. Needs to be the same algorithm
+// amongst all implementations.
+fn stretch_key(key: &SigningKey, result: &mut [u8]) {
```
Comment (rphmeier) [inline libp2p-secio/src/handshake.rs]: seems like something that should be tested (just skimming)

**Comment by tomaka** [comment]
Just realized I had committed some non-secio-related changes. I force-pushed to remove them.

File: libp2p-secio/regen_structs_proto.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+sudo docker run --rm -v `pwd`:/usr/code:z -w /usr/code rust /bin/bash -c " \
+    apt-get update; \
+    apt-get install -y protobuf-compiler; \
+    cargo install protobuf; \
+    protoc --rust_out . structs.proto; \
+    protoc --rust_out . keys.proto"
+
+mv -f structs.rs ./src/structs_proto.rs
+mv -f keys.rs ./src/keys_proto.rs
```
Comment (debris) [inline libp2p-secio/regen_structs_proto.sh]: 🎉 we should use the same snippet in parity repo!

File: libp2p-secio/src/codec/encode.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	type SinkItem = BytesMut;
+	type SinkError = S::SinkError;
+
+	fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+		let capacity = item.len() + self.hmac_key.digest_algorithm().output_len;
+
+		let mut out_buffer = BytesMut::with_capacity(capacity);
+		// Note: Alternatively to `extend`, we could also call `advance_mut()`, which will add
+		//		 uninitialized bytes to the buffer. But that's unsafe.
+		out_buffer.extend((0..item.len()).map(|_| 0));
```
Comment (debris) [inline libp2p-secio/src/codec/encode.rs]: imo, repeat take looks better ;)

```rust
out_buffer.extend(repeat(0).take(item.len());
```

File: libp2p-secio/src/handshake.rs [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+		local_tmp_pub_key: [0; agreement::PUBLIC_KEY_MAX_LEN],
+	};
+
+	// The handshake messages all start with a 4-bytes message length prefix.
+	let socket =
+		length_delimited::Builder::new().big_endian().length_field_length(4).new_framed(socket);
+
+	let future = future::ok::<_, SecioError>(context)
+        // Generate our nonce.
+        .and_then(|mut context| {
```
Comment (debris) [inline libp2p-secio/src/handshake.rs]: bad indentation, use tabs

**Review by debris — COMMENTED** [review]
neat!

File: libp2p-secio/src/handshake.rs [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+
+	#[test]
+	fn stretch() {
+		let mut output = [0u8; 32];
+
+		let key1 = SigningKey::new(&SHA256, &[]);
+		stretch_key(&key1, &mut output);
+		assert_eq!(
+			&output,
+			&[
```
Comment (tomaka) [inline libp2p-secio/src/handshake.rs]: Sorry for the very bad formatting, but I think [the option to fix that in rustfmt doesn't work](https://github.com/rust-lang-nursery/rustfmt/blob/master/Configurations.md#array_horizontal_layout_threshold).

`rustfmt` also likes to convert tabs into spaces for some reason, but only at some specific places.

File: libp2p-secio/src/codec/encode.rs (hunk 3 above)
Comment (rphmeier) [reply libp2p-secio/src/codec/encode.rs]: requires an allocation in this case though

Comment (kirushik) [reply libp2p-secio/src/codec/encode.rs]: @rphmeier I've updated [the bench](https://gist.github.com/kirushik/e0d93759b0cd102f814408595c20a9d0) with a third variant -- allocate the whole structure as a Vec of zeroes, then use `split_at_mut` to get two mutable slices into it.
It's an additional 3× improvement, and no extra memory copying, as far as I can tell.

**Pull request merged**